### ISSUE TITLE
clicking hash link recorded in history twice

### ISSUE THREAD

**Issue by kyontan** [sveltejs/kit#3636]
### Describe the bug

When opening `#anchor-a` then click a link to `#anchor-b`, history **newly** records entries `#anchor-a` and `#anchor-b` though `#anchor-a` is previous location.

Therefore, hisotry looks like:

```
#anchor-a (initial view)
#anchor-b
#anchor-b (when clicked #anchor-b)
#anchor-c
#anchor-c (when clicked #anchor-c)
#anchor-d
#anchor-d (when clicked #anchor-d)
...
```

Disabling `router` by `export const router = false;` mitigate this issue, so that I think router has some issue around this.

### Reproduction

https://stackblitz.com/edit/sveltekit-ivmzya?devtoolsheight=33&file=src/routes/index.svelte

Easily understand by following steps:

1.  `Open in New Window` in Stackblitz
2.  Clicking anchor links some times
3.  Check a browser history, looks like:

<img width="319" alt="スクリーンショット 2022-01-31 3 59 07" src="https://user-images.githubusercontent.com/802339/151713646-85bb8908-f253-40de-9879-d6cb080fadeb.png">

(Also I notified `null` is oddly appealing on first clicking `#anchor-b`)

### System Info

```shell
# npx envinfo --system --binaries --browsers --npmPackages "{svelte,@sveltejs/*,vite}"
  System:
    OS: macOS 11.4
    CPU: (8) x64 Apple M1
    Memory: 24.88 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.13.2 - ~/.nodebrew/current/bin/node
    Yarn: 1.22.17 - ~/.nodebrew/current/bin/yarn
    npm: 8.1.2 - ~/.nodebrew/current/bin/npm
  Browsers:
    Firefox: 95.0
    Safari: 14.1.1
  npmPackages:
    @sveltejs/adapter-auto: next => 1.0.0-next.17
    @sveltejs/kit: next => 1.0.0-next.251
    svelte: ^3.46.3 => 3.46.3
```
```


### Severity

blocking an upgrade

### Additional Information

_No response_

**Comment by kyontan** [comment]
I think replacing `<a href="#anchor">...</a>` with `<a on:click={() => goto("#anchor")}>...</a>` could workaround this.

**Comment by Glench** [comment]
Maybe related to #3621?

**Comment by bluwy** [comment]
Yeah it does look like that this issue causes #3621. cc @vicky1999 in case you're already working on a fix.

**Comment by vicky1999** [comment]
@bluwy I think the above PR will solve the issue.  but, we need to check this once.

**Comment by aolose** [comment]
I think some code change the history ,so I global searched `history.`  then found  it at:
kit/src/runtime/client/router.js line:191

```js
// Check if new url only differs by hash
if (url.href.split('#')[0] === location.href.split('#')[0]) {
	// Call `pushState` to add url to history so going back works.
	// Also make a delay, otherwise the browser default behaviour would not kick in
	setTimeout(() => history.pushState({}, '', url.href));
	const info = this.parse(url);
	if (info) {
		return this.renderer.update(info, [], false);
	}
	return;
}
```

I don't know what does these code do, 
the `history.pushState` here will record history one more time.
I try to delete it and run the Reproduction ,then hash links works fine~

**Comment by benmccann** [comment]
I've sent a PR for a failing test that can be uncommented when this issue is fixed: https://github.com/sveltejs/kit/pull/3880

**Comment by benmccann** [comment]
This was fixed recently. The test has now been uncommented and passes:

https://github.com/sveltejs/kit/blob/a37f3b3de07304641cdae5d3686bfda6861b4489/packages/kit/test/apps/basics/test/test.js#L217

**Comment by kyontan** [comment]
Great thanks to @benmccann !
